system/settings/settings.c

```diff
@@ -0,0 +1,1717 @@
+ * Name: settings_setstorage
+ *
+ * Description:
+ *    Sets a file to be used as a settings storage.
+ *    Except from the first file, if loading the file causes any changes
+ *    to the settings, then the new map will be dumped to all files
+ *    (effectively it syncs all storages).
+ *
+ * Input Parameters:
+ *    file             - the filename of the storage to use
+ *    type             - the type of the storage (BINARY or TEXT)
+ *
+ * Returned Value:
+ *   Success or negated failure code
+ *
+ ****************************************************************************/
+
+int settings_setstorage(FAR char *file, enum storage_type_e type)
+{
+  FAR storage_t *storage = NULL;
+  int ret = OK;
+  int idx = 0;
+  uint32_t h;
+
+  ret = pthread_mutex_lock(&g_settings.mtx);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  if (!g_settings.initialized)
+    {
+      ret = -EACCES;
+      goto errout;
+    }
+
+  while ((idx < CONFIG_SYSTEM_SETTINGS_MAX_STORAGES) &&
+         (g_settings.store[idx].file[0] != '\0'))
+    {
+      idx++;
+    }
+
+  DEBUGASSERT(idx < CONFIG_SYSTEM_SETTINGS_MAX_STORAGES);
+  if (idx >= CONFIG_SYSTEM_SETTINGS_MAX_STORAGES)
+    {
+      ret = -ENOSPC;
+      goto errout;
+    }
+
+  DEBUGASSERT(strlen(file) < CONFIG_SYSTEM_SETTINGS_MAX_FILENAME);
+  if (strlen(file) >= CONFIG_SYSTEM_SETTINGS_MAX_FILENAME)
+    {
+      ret = -EINVAL;
+      goto errout;
+    }
+
+  storage = &g_settings.store[idx];
+  strncpy(storage->file, file, sizeof(storage->file));
+  storage->file[sizeof(storage->file) - 1] = '\0';
+
+  switch (type)
+  {
+    case STORAGE_BINARY:
+      {
+        storage->load_fn = load_bin;
+        storage->save_fn = save_bin;
+        storage->size_fn = size_bin;
+      }
+      break;
+
+    case STORAGE_TEXT:
+      {
+        storage->load_fn = load_text;
+        storage->save_fn = save_text;
+        storage->size_fn = size_text;
+      }
+      break;
+
+    case STORAGE_EEPROM:
+      {
+        storage->load_fn = load_eeprom;
+        storage->save_fn = save_eeprom;
+        storage->size_fn = size_eeprom;
+      }
+      break;
+
+    default:
+      {
+        DEBUGASSERT(0);
+      }
+      break;
+  }
+
+  ret = storage->load_fn(storage->file);
+
+  h = hash_calc();
+
+  /* Only save if there are more than 1 storages. */
+
+  if ((storage != &g_settings.store[0]) && ((h != g_settings.hash) ||
+      (access(file, F_OK) != 0)))
+    {
+      signotify();
+      save();
+    }
+
+  g_settings.hash = h;
+
+errout:
+  pthread_mutex_unlock(&g_settings.mtx);
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: settings_sync
+ *
+ * Description:
+ *    Synchronizes the storage.
+ *
+ * Input Parameters:
+ *    none
+ *
+ * Returned Value:
+ *    Success or negated failure code
+ *
+ ****************************************************************************/
+
+int settings_sync(void)
```

Comment:
A better way to solve your issues would be to:
* Completelly ditch `settings_savepending()`.
* Add a new argument to `settings_sync()` whether it should just schedule the sync, or also wait for the cache to be dumped.
* If a wait for the cache is specified, `settings_sync()` can wait just as `settings_clear()` does.
* Before exiting your application (or task), call  `settings_sync()`.
* In this case, `settings_sync()` can also call `syncfs()` internally.

Reply:
Yes, that is neater and is why we emailed about it. I will rework/check.

Reply:
done